### ISSUE TITLE
[c++20] Clang doesn't show an error for defaulted comparison operators with mixed parameters

### ISSUE THREAD

**Issue by jensmassberg** [llvm/llvm-project#62880]
By the standard both parameters of a defaulted comparison operator for a class `C` should be either of type `const C&` or type `C` but not mixed, see https://eel.is/c++draft/class.compare.default#1.

However, Clang accepts the following example which is declined by GCC and MSCV:

```c++
#include <compare>

template <class C>
struct Bar {
  C i;
  friend bool operator==(const Bar&, Bar) = default;
};
```

See https://gcc.godbolt.org/z/M7Y9qGrsE

Note that an error is printed if it is class without a template https://gcc.godbolt.org/z/4sb5hMn3M

**Comment by jensmassberg** [comment]
Additionally Clang should error out if the comparison operator has a single parameter that isn't a const reference, see https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p2002r1.html

Example:
```
#include <compare>

template <class C>
struct Bar {
  C i;
  bool operator!=(Bar) const = default;
};
```

The code is accepted by Clang, but rejected by GCC and MSVC:
https://gcc.godbolt.org/z/58Y4MGxGa

**Comment by shafik** [comment]
CC @royjacobson this feels like you area
